### PR TITLE
UDP port binding + virtualization between apps and capsules

### PR DESCRIPTION
**Pull request by hudson-ayers**
### Pull Request Overview

This pull request adds support for UDP virtualization between both in-kernel capsules and userspace apps. It introduces a port binding interface that relies on an in-kernel table of bound ports to enforce binding for a limited number of in-kernel capsules. It also introduces in-kernel queuing of UDP packets for transmission and virtualizes UDP reception between capsules and applications. It includes several comprehensive tests of UDP sending, receiving, and virtualization. It also includes small fixes for the networking stack discovered throughout the course of developing this PR.

Port binding for apps is still managed within the UDP driver, but the UDP driver and this in-kernel table have references to each other that allow them to prevent capsules and applications from binding to the same port. Binding is enforced via a FCFS model, with the caveat that capsules will always get first dibs on binding because their initialization code is executed prior to any applications being executed.

To enable testing of virtualization between capsules and apps, this PR should be tested alongside the corresponding PR in libtock-c which provides several userspace tests that can be paired with the in-kernel apps introduced in this PR.

This PR also introduces the "Buffer" struct that Amit proposed awhile back, and modifies the networking stack above the sixlowpan layer to use it almost exclusively on the send side. It makes it possible to pass "slices" of larger buffers down the stack without making it impossible to recover the original, full buffer, and without requiring passing a length alongside every call that passes a packet buffer.

This PR was authored by myself and Armin (@ajnam12)

### Testing Strategy

This PR was tested using the in-kernel tests introduced in this PR, which involve sending and receiving packets between two Imixes in a variety of scenarios, and testing various edge cases of port binding.

### Documentation Updated

Documentation in the relevant files has been updated, but the network stack document in docs/ still needs to reflect these changes.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

File: kernel/src/debug.rs [hunk 1]
```diff
@@ -219,8 +219,8 @@ pub struct DebugWriter {
 /// needed so the debug!() macros have a reference to the object to use.
 static mut DEBUG_WRITER: Option<&'static mut DebugWriterWrapper> = None;
 
-pub static mut OUTPUT_BUF: [u8; 64] = [0; 64];
-pub static mut INTERNAL_BUF: [u8; 1024] = [0; 1024];
+pub static mut OUTPUT_BUF: [u8; 512] = [0; 512];
+pub static mut INTERNAL_BUF: [u8; 4096] = [0; 4096];
```
Comment (gendx) [inline kernel/src/debug.rs]: This seems unrelated to this change.

Comment (hudson-ayers) [reply kernel/src/debug.rs]: You're right, thanks, fixed

File: kernel/Cargo.toml [hunk 2]
```diff
@@ -7,3 +7,6 @@ edition = "2018"
 [dependencies]
 tock-registers = { path = "../libraries/tock-register-interface" }
 tock-cells = { path = "../libraries/tock-cells" }
+#line below for "pub in" doesn't work b/c cyclic dependency
+#capsules = { path = "../capsules" }
+
```
Comment (bradjc) [inline kernel/Cargo.toml]: Remove

File: kernel/src/lib.rs [hunk 3]
```diff
@@ -21,6 +21,8 @@ pub mod hil;
 pub mod introspection;
 pub mod ipc;
 pub mod syscall;
+#[macro_use]
```
Comment (bradjc) [inline kernel/src/lib.rs]: I don't think we need this.

File: kernel/src/udp_port_table.rs [hunk 4]
```diff
@@ -0,0 +1,249 @@
+//! UDP port table implementation for enforcing port binding. This table is
+//! checked when packets are sent/received by a UDPSendStruct/UDPReceiver.
+//! UdpPortBinding provides an opaque descriptor object that allows the holder
+//! interact with the bound port table. Only the holder of the UdpPortBinding
+//! object can interact with its own corresponding location in the bound port
+//! table. In order to bind to a particular port as sending/receiving, one must
+//! obtain the corresponding sender/receiving binding from UdpPortBinding.
```
Comment (bradjc) [inline kernel/src/udp_port_table.rs]: I'm not sure I understand what exactly the code in this file has to do. I think a higher-level description would be useful, as well as a description of what is stored here vs. what is stored elsewhere in a capsule.

File: kernel/src/udp_port_table.rs [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+#[derive(Clone, Copy, PartialEq)]
+pub enum PortEntry {
+    Port(u16),
+    Unbound,
+}
+
+// We need Option<PortEntry> to distinguish between the case in which we have
+// a UdpPortSocket that is not bound to a port and an index where there is no
+// UdpPortSocket allocated.
+static mut PORT_TABLE: [Option<PortEntry>; MAX_NUM_BOUND_PORTS] = [None; MAX_NUM_BOUND_PORTS];
```
Comment (bradjc) [inline kernel/src/udp_port_table.rs]: I'm not sure I understand this data structure. What do the entries correspond to?

File: kernel/src/udp_port_table.rs [hunk 6]
```diff
@@ -0,0 +1,249 @@
+//! UDP port table implementation for enforcing port binding. This table is
```
Comment (bradjc) [inline kernel/src/udp_port_table.rs]: Blank line after a short header.

Also I think creating a `net` folder to mirror the capsules would be useful.

File: kernel/src/udp_port_table.rs [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+    // Mark new as unsafe so that the port table is only generated in trusted
+    // code.
+    pub unsafe fn new() -> UdpPortTable {
+        UdpPortTable {
+            port_array: TakeCell::new(&mut PORT_TABLE),
+            user_ports: OptionalCell::empty(),
+        }
+    }
+
+    pub unsafe fn set_user_ports(&self, user_ports_ref: &'static dyn PortQuery) {
```
Comment (bradjc) [inline kernel/src/udp_port_table.rs]: This looks like a case where it shouldn't be unsafe, but should rely on capabilities.

File: capsules/src/net/buffer.rs [hunk 8]
```diff
@@ -0,0 +1,151 @@
+//! Defines a Buffer type which can be used to pass a section of a larger
+//! buffer but still get the entire buffer back in a callback
+//!
+//! Author: Amit Levy
```
[the remark this answers is not in the document]
Comment (hudson-ayers) [reply capsules/src/net/buffer.rs]: It seems like it would be better suited in libraries/ ? I view at as similar to the stuff in tock-cells or enum_primitive

File: kernel/src/net/udp_port_table.rs [hunk 9]
```diff
@@ -0,0 +1,293 @@
+//! In-kernel structure for tracking UDP ports bound by capsules.
```
Comment (bradjc) [inline kernel/src/net/udp_port_table.rs]: Foreword: if this table is going to be in the kernel, then I think it is worth making sure we are happy with the interface and implementation, and that could potentially take some discussion.

With that disclaimer, my first question is why is this in the kernel (I use "kernel" to mean "kernel crate")? I ask for two reasons: 1) the code that enforces that the port bindings are valid is not in the kernel, and 2) ports provided to userspace are not checked by this code. Given both of those, why not just leave port management up to the capsules in `capsules/net`? Port management certainly could be a kernel level operation, but for that to be really effective it then also has to enforce the bindings (i.e. route incoming packets based on the bindings).

If this should be in the kernel, then why does it not manage userland ports as well? Perhaps something could request a range of ports for userland. I do sympathize with the complexity of writing code where both capsules and userland processes share resources, however. I don't think it is very natural at the moment, and I don't think we have a standard way to do it.

If this should be in the kernel, then I think changing some of the names would be worthwhile.
-`PORT_TABLE` => `USED_KERNEL_PORTS`
-`UdpPortSocket` => `UdpSocket`
-`UdpPortTable` => `UdpPortManager`
-`PortEntry` => `SocketBindingEntry`
-`UdpReceiverBinding` => `UdpPortBindingReceive`

Also hopefully others will add comments as well...

Comment (hudson-ayers) [reply kernel/src/net/udp_port_table.rs]: Thanks for taking time to go through all of this!

Thinking about it now, I think it would be possible to move this into `capsules/net/udp`. Originally this was all in the kernel because we were keeping the door open  toward moving some portion of sending/receiving UDP packets into the kernel such that bindings could be managed and enforced in the kernel, and other potential network security considerations could be enforced in the kernel. 

In hindsight, it seems unlikely that we will move enforcement into the kernel anytime soon. Given this, I think having `UdpPortTable::new()` require some capability to be called would have basically the same effect as our current implementation if `udp_port_table.rs` was in `capsules/net/udp`. This is a reasonable approach for now, I'll make that change.

Re: why does this not manage userland ports as well?
A: There are several advantages to storing userland ports in the grant region of each app. First, grant regions are dynamically allocated / removed when apps are added / removed, so storing userland ports in the grant region scales to any number of apps without the kernel having to guess as to how many userland apps will bind to ports. Second, this approach ensures that whenever an app is removed from a board any ports it has bound are immediately unbound, as the removal of the grant region effectively removes the binding -- enforcing this otherwise is difficult, as `drop()` is not called on objects in the grant region because we simply zero out that memory when apps are removed. This second point is important, as we don't want a design where an app failing locks a replacement of that app out from a particular port it wanted to use. Third, the current design relies on each capsule holding a single `UdpSender` / `UdpReceiver` for each port it binds to, and dynamically allocating those objects for each new app sounds difficult. The two-tier virtualization in this PR also makes it easy to prioritize capsules over userspace apps when it comes to sending queued packets.

Even though I am going to move the port table to capsules, I'll still apply the name changes, they are definitely an improvement.

File: capsules/src/net/buffer.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (hudson-ayers) [reply capsules/src/net/buffer.rs]: After reading through the README for `libraries/` I changed my mind, as this has not necessarily "become reasonably stable" at this point, so I moved it to kernel instead

File: boards/imix/src/main.rs [hunk 10]
```diff
@@ -403,18 +410,32 @@ pub unsafe fn reset_handler() {
...
+        udp_send_mux,
+        udp_recv_mux,
+        udp_port_table,
+        local_ip_ifaces,
+    )
+    .finalize(());
+
+    // Only include to run kernel tests, do not include during normal operation
+    let udp_lowpan_test =
+        udp_lowpan_test::initialize_all(udp_send_mux, udp_recv_mux, udp_port_table, mux_alarm);
```
Comment (bradjc) [inline boards/imix/src/main.rs]: Presumably this shouldn't be merged?

File: boards/imix/src/main.rs [hunk 11]
```diff
@@ -466,6 +487,8 @@ pub unsafe fn reset_handler() {
 
     debug!("Initialization complete. Entering main loop");
 
+    udp_lowpan_test.start();
```
Comment (bradjc) [inline boards/imix/src/main.rs]: Or maybe these should just be commented out?

File: kernel/src/common/buffer.rs [hunk 12]
```diff
@@ -0,0 +1,151 @@
+//! Defines a Buffer type which can be used to pass a section of a larger
+//! buffer but still get the entire buffer back in a callback
+//!
+//! Author: Amit Levy
+
+use core::ops::{Bound, Range, RangeBounds};
+use core::ops::{Index, IndexMut};
+use core::slice::SliceIndex;
+
+pub struct Buffer<'a, T> {
```
Comment (bradjc) [inline kernel/src/common/buffer.rs]: Maybe `SliceableBuffer` as a name? Or `LeaseableBuffer`? In any case I think we need something that signals this is not an ordinary buffer.

Comment (hudson-ayers) [reply kernel/src/common/buffer.rs]: `LeasableBuffer` sounds good to me, what should I do with the commented out tests in this file? I could just delete them, but they are sort of useful to have around in case anyone ever wanted to modify this interface and check that it still works.

File: kernel/src/common/mod.rs [hunk 13]
```diff
@@ -15,6 +15,7 @@ pub mod registers {
     pub use tock_registers::{register_bitfields, register_structs};
 }
 
+pub mod buffer;
```
Comment (bradjc) [inline kernel/src/common/mod.rs]:
```suggestion
pub mod leasable_buffer;
```

File: kernel/src/common/buffer.rs [hunk 14]
```diff
@@ -0,0 +1,129 @@
+//! Defines a LeasableBuffer type which can be used to pass a section of a larger
+//! buffer but still get the entire buffer back in a callback
+//!
+//! Author: Amit Levy
+//!
+//! Usage
+//! -----
+//!
+//! `slice()` is used to set the portion of the `LeasableBuffer` that is accessbile.
+//! `reset()` makes the entire `LeasableBuffer` accessible again.
```
Comment (bradjc) [inline kernel/src/common/buffer.rs]: Does anything prevent the lower layer from calling `reset()`?

Comment (hudson-ayers) [reply kernel/src/common/buffer.rs]: Currently, no. This design is more to protect against accidental misuse of the whole buffer. Prior to using this construct we would always pass the entire buffer along with the length that the lower layer was actually supposed to access, but this was more mistake prone

**Comment by hudson-ayers** [comment]
Travis is failing because it is trying to compile my example usage code in the doc comment? What am I doing wrong that that is happening

File: kernel/src/common/buffer.rs (hunk 14 above)
Comment (bradjc) [reply kernel/src/common/buffer.rs]: Ok gotcha. I'm wondering if `slice()` could return something that must be passed to `reset()` and that can't be created by the thing that got the sliced buffer.

Comment (bradjc) [reply kernel/src/common/buffer.rs]: But that wouldn't be in this PR.

**Comment by bradjc** [comment]
Marking this one last-call to raise attention, and will merge in a couple days if there are no objections.

File: capsules/src/net/ipv6/ipv6.rs [hunk 15]
```diff
@@ -301,6 +303,7 @@ impl IP6Header {
 /// This defines the currently supported `TransportHeader` types. The contents
 /// of each header is encapsulated by the enum type. Note that this definition
 /// of `TransportHeader`s means that recursive headers are not supported.
+#[derive(Copy, Clone)] // TODO: is this ok?
```
Comment (ppannuto) [inline capsules/src/net/ipv6/ipv6.rs]: Is this TODO still an open question? (fine if so, but doesn't compile without it -- if there's an underlying buffer / shallow-copy / etc concern behind this comment, might be nice to document that)

File: capsules/src/net/ipv6/ipv6.rs [hunk 16]
```diff
@@ -343,11 +346,18 @@ impl IPPayload<'a> {
...
     /// (when serialized)
-    pub fn set_payload(&mut self, transport_header: TransportHeader, payload: &[u8]) -> (u8, u16) {
+    pub fn set_payload(
+        &mut self,
+        transport_header: TransportHeader,
+        payload: &LeasableBuffer<'static, u8>,
+    ) -> (u8, u16) {
         if self.payload.len() < payload.len() {
-            // TODO: Error
+            debug!("Error in set_payload for ipv6_packet");
```
Comment (ppannuto) [inline capsules/src/net/ipv6/ipv6.rs]: If I understand correctly, two lines later this is going to `panic!` when `self.payload[i]` goes beyond valid array bounds -- maybe this should return here? Or is the goal to leave the `panic!` in for now until better error-handling bubbles through the rest of the networking stack?

File: capsules/src/net/udp/driver.rs [hunk 17]
```diff
@@ -6,18 +6,22 @@
 //! Also exposes a list of interface addresses to the application (currently
 //! hard-coded).
 
+/// Syscall number
```
Comment (ppannuto) [inline capsules/src/net/udp/driver.rs]: This is a bit of a weird comment -- we don't usually document import purposes, particularly not with doc comments

File: capsules/src/net/ipv6/ipv6.rs (hunk 16 above)
Comment (hudson-ayers) [reply capsules/src/net/ipv6/ipv6.rs]: Goal for now is to leave panic!(), the whole networking stack definitely needs improved error handling eventually but don't want to worry about that in this PR.

File: capsules/src/net/udp/driver.rs (hunk 17 above)
Comment (hudson-ayers) [reply capsules/src/net/udp/driver.rs]: I think that is leftover from the file being copied from the 15.4 driver, but it looks like the same doc comment appears in the usb_user driver as well, I will remove all 3.

File: capsules/src/net/ipv6/ipv6.rs (hunk 15 above)
Comment (hudson-ayers) [reply capsules/src/net/ipv6/ipv6.rs]: Good catch, for now I think the overhead of this copy is fine, the largest header possible is a 40 byte TCP header which isn't huge and currently it looks like only 2 potentially unnecessary copies of a TransportHeader occur in the process of sending a packet, and changing to a design without copy would require a decent amount of restructuring.
